File: genetic-web/src/main/java/genetic/web/controller/PackerController.java
package genetic.web.controller;

import genetic.packer.dto.request.ParamsDto;
import genetic.packer.dto.request.RequestDto;
import genetic.packer.dto.response.ResponseDto;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * @author piotr.larysz
 */
@RestController
@RequestMapping("packer")
public class PackerController {

    @Autowired
    private BiFunction<RequestDto, ParamsDto, ResponseDto> requestResponseAdapter;

    @RequestMapping(method = RequestMethod.POST)
    public ResponseDto pack(@Valid @RequestBody RequestDto request, ParamsDto params) {
        return requestResponseAdapter.apply(request, params);
    }
}

File: genetic-web/src/main/java/genetic/web/controller/PackerController.kt
package genetic.web.controller

import genetic.packer.dto.request.EmbryoDto
import genetic.packer.dto.request.ParamsDto
import genetic.packer.dto.request.RequestDto
import genetic.packer.dto.response.ResponseDto
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.validation.Valid

@RestController
@RequestMapping("packer")
class PackerController(
    private val requestResponseAdapter: (RequestDto<EmbryoDto>, ParamsDto) -> (ResponseDto)
) {

  @PostMapping
  fun pack(@RequestBody request: @Valid RequestDto<EmbryoDto>, params: ParamsDto): ResponseDto {
    return requestResponseAdapter(request, params)
  }
}